appliances/Harbor/appliance.sh

```diff
@@ -0,0 +1,304 @@
+    if [ $? -ne 0 ]; then
+	    msg error "Harbor installation script failed, aborting..."
+        exit 1
+    else
+        msg info "Harbor installation script finished successfully. Waiting 60s before final health checks..."
+    fi
+    sleep 60
+    msg info "Running final health checks..."
+    wait_for_docker_containers
+    cleanup_installation
+    msg info "Bootstrap phase finished"
+
+}
+
+### Function Definitions
+
+check_internet_access() {
+    # Ping Google's public DNS server
+    if ping -c 1 8.8.8.8 &> /dev/null; then
+        msg info "Internet access OK"
+        return 0
+    else
+        msg error "The VM does not have internet access. Aborting Harbor deployment..."
+        exit 1
+    fi
+}
+
+install_requirements(){
+    apt update && apt install openssl ca-certificates curl gnupg docker.io docker-compose -y
+    if [ $? -eq 0 ]; then
+	    msg info "Dependencies installation finished successfully. Running extra checks..."
+    else
+        msg error "Dependencies installation failed. Aborting..."
+    fi
+    check_docker
+    check_docker_compose
+}
+
+check_docker() {
+    msg info "Checking if Docker is installed and running..."
+    if ! command -v docker &> /dev/null; then
+        msg error "Docker could not be found, Docker installation failed..."
+    else
+        msg info "Docker is installed, ensuring it is running..."
+        if ! systemctl is-active --quiet docker; then
+            msg warning "Docker is not running, attempting to start Docker..."
+            systemctl start docker
+            systemctl enable docker
+            sleep 5
+        fi
+        if ! systemctl is-active --quiet docker; then
+            msg error "Failed to start Docker"
+            exit 1
+        else
+            msg info "Docker is running"
+            return 0
+        fi
+    fi
+}
+
+check_docker_compose() {
+    msg info "Checking if docker-compose is installed..."
+    if ! command -v docker-compose &> /dev/null; then
+        msg error "docker-compose could not be found, aborting..."
+        exit 1
+    else
+        msg info "docker-compose is installed"
+        return 0
+    fi
+}
+
+
+
+generate_ssl_certs() {
+    if [ -z "${HARBOR_HOSTNAME}" ]; then
+        msg info "HARBOR_HOSTNAME is not set or is empty. Generating certs with 'example.com'"
+        HARBOR_HOSTNAME=example.com
+    fi
+    openssl req -new -newkey rsa:2048 -sha256 -days 365 -nodes -x509 \
+    -keyout /root/certs/server.key -out /root/certs/server.crt \
+    -subj "/C=US/ST=State/L=City/O=Organization/CN=$HARBOR_HOSTNAME"
+    if [ $? -eq 0 ]; then
+        msg info "Certificate generated successfully"
+    else
+        msg info "Error generating SSL certificate. Resuming..."
+    fi
+}
+
+
+download_unpack_harbor() {
+    msg info "Downloading Harbor..."
+    url=$(curl -s https://api.github.com/repos/goharbor/harbor/releases/latest | grep browser_download_url | cut -d '"' -f 4 | grep '\.tgz$' | grep online)
```

Comment:
Have you considered `set -e` + `curl -f` for some better error handling?

Reply:
Hmmm didn't think about it. There are some cases where I want to resume the config even after a non-zero exit code (e.g. line 186). It shouldn't be convenient to run `set -eo pipefail` for that case, not sure if there's a way to manage it

Reply:
So the idea is to use `PIPESTATUS` at some point ($? returns only the error code from the last command in the pipeline), right? 

> It shouldn't be convenient to run set -eo pipefail for that case, not sure if there's a way to manage it

So actually instead of a single if statement and a big pipeline, it's possible to split it into smaller pipelines and more if statements, then you don't have to check specific error codes I guess. :thinking:

Reply:
Nice, I'll think about it. Thanks sir